### PR TITLE
fix(c4): break rows on c4ShapeInRow instead of the display width

### PR DESCRIPTION
**Pull request by filipsajdak**
C4 row breaking depends on the size of the monitor the diagram is rendered on. This removes that dependency, so a C4 diagram lays out the same way everywhere.

Found while answering @pbrolin47's question on #7842 about a C4 layout difference in the 11.17.0 release-candidate image tests that he could not reproduce on a local dev server ([diagnosis](https://github.com/mermaid-js/mermaid/pull/7842#issuecomment-5269085434), [measured numbers](https://github.com/mermaid-js/mermaid/pull/7842#issuecomment-5269201184)). This is the cause of the "cannot reproduce locally" half of that report. The width change he was looking at is separate and expected - it is #7842's reviewed change.

## The defect

`c4Renderer.ts` seeded the row budget from the display:

```ts
screenBounds.data.widthLimit = screen.availWidth;
```

That was the only place in `packages/mermaid/src` reading the physical display, so C4 was the only diagram type in mermaid whose geometry depended on the machine. A row ended when the next element would cross that budget, so with default config (`width` 216, `c4ShapeMargin` 50, `diagramMarginX` 50) the fourth element of a row lands at `450 + 4w` and the row breaks once `w >= (availWidth - 450) / 4`:

| `screen.availWidth` | where it comes from | elements per row at the default width |
| --- | --- | --- |
| 800 | headless Chrome (the `e2e` job) - Chrome's Linux new-headless hard-codes an 800x600 virtual screen | 2 |
| 1280 | Electron + Cypress's Xvfb (`-screen 0 1280x1024x24`, the Applitools job) | 3 |
| 1512-2560 | a developer's monitor | 4 |
| 0 | jsdom (`Screen-impl.js` hard-codes `availWidth = 0`) | 1 - a single column |

Nesting made it worse: the budget was divided by `min(c4BoundaryInRow, siblings)` at every level, so a boundary two levels deep got a quarter of it and its elements stacked one per row. And because the test compared an absolute x coordinate against a width budget, the effective threshold varied with nesting depth.

One user-visible consequence: `UpdateLayoutConfig($c4ShapeInRow="6")` was silently capped, because the 5th element already crossed the budget on most displays.

## The change

A row ends after `c4ShapeInRow` elements, however wide they measure:

```ts
if (this.nextData.cnt > c4ShapeInRow) {
```

`c4ShapeInRow` already existed as the count-based break, so this removes the pixel test rather than replacing it. Row topology is now a function of the diagram source and its config alone. Element widths still depend on the font, but only box sizes vary now - not which row an element lands in.

`widthLimit` had a second, non-obvious role: it was also the wrap budget for boundary label/type/description text. That is now the widest row the grid can produce (`c4ShapeInRow` elements plus their margins), so boundary text wrapping stops varying by machine too, instead of silently losing wrapping.

## Evidence

`c4Renderer.spec.ts` renders through the public API with `getBBox` stubbed, so every element self-sizes to the `c4.width` floor and the grid coordinates are exact. Four cases: a row fills to `c4ShapeInRow` before the next starts; the same source places identically whatever the display reports; `$c4ShapeInRow="6"` is honoured; a boundary's elements stay on one row.

All four are red on `develop` and green here. On `develop` jsdom reports `availWidth = 0`, so the five elements stack in a single column at x 258 (y 403, 643, 883, 1123, 1363) instead of laying out 4 + 1. Worth knowing on its own: **no unit test could previously exercise C4 layout at all**, because every element landed on its own row regardless of the diagram.

The 130 existing C4 unit tests still pass; they are parser and db tests and never asserted geometry.

`eslint.config.js` gains `no-restricted-globals` for `screen` under `packages/mermaid/src`, so this cannot regress. Verified it fires by re-adding a `screen.availWidth` read.

## This moves screenshots, deliberately

Argos will show a C4 diff, and it is the point of the change rather than a side effect. Measured across all 25 C4 fixtures, `develop` versus this branch:

| `screen.availWidth` | fixtures rendering differently |
| --- | --- |
| 800 (headless Chrome, the `e2e` job) | 4 of 25 |
| 1280 (Xvfb) | 1 of 25 |
| 1512 (a laptop display) | 0 of 25 |

Where a narrow CI display previously forced fewer elements per row, a row now holds up to `c4ShapeInRow`, so those diagrams get wider and shorter. On a developer's display the current fixtures do not move at all, which is exactly why this defect went unnoticed.

The same measurement states the defect most directly: on `develop`, four fixtures render differently depending only on the width of the monitor, and on this branch the count is 0 at every width. The changeset says so.

Flagging the timing explicitly: 11.17.0 is being prepared right now, and this alters the C4 baseline it is cut from. @pbrolin47 - happy to hold this until the release is out if that is easier; the defect is not new and is not getting worse.

## Not fixed here

- The absolute-x versus width-budget comparison is gone with the pixel test, but the surrounding grid is otherwise untouched.
- Boundaries still wrap every `c4BoundaryInRow` by index, unchanged.
- `c4.wrap` is still ignored in favour of the root `wrap` (#7949).
- Element widths remain font-dependent, so box sizes still differ between CI and a developer machine - only the row structure is now stable.

All of this disappears with #8042, which replaces this grid with the unified layout pipeline and deletes `c4Renderer.ts`. This is deliberately a small change to a file with a limited remaining life, chosen over re-designing the row-break rule.

## Summary

- Made C4 row placement deterministic by using `c4ShapeInRow` instead of display width.
- Derived boundary wrapping limits from the grid’s maximum row width.
- Added an ESLint rule that prevents `screen`-based layout calculations in Mermaid source.
- Added renderer tests for row capacity, display independence, custom row limits, and boundary layout.
- Added a changeset documenting the behavior change.

## Testing

- Existing C4 tests pass.
- Screenshot differences are expected in narrow display environments.

### CODE REVIEW

File: packages/mermaid/src/diagrams/c4/c4Renderer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        _window.Element.prototype.getBBox = () => ({ x: 0, y: 0, width: 100, height: 100 });
+        _window.Element.prototype.getComputedTextLength = () => 50;
+      },
+    });
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (global as any).window = dom.window;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (global as any).document = dom.window.document;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    (global as any).MutationObserver = undefined;
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply packages/mermaid/src/diagrams/c4/c4Renderer.spec.ts:34]: Both findings taken. The second one was right about the gap and wrong about the fix, and the wrong part would have broken the lint job, so it is worth spelling out.

## `MutationObserver` not restored - fixed

Correct, and it was asymmetric with the two globals either side of it: `window` and `document` were saved and restored, `MutationObserver` was set to `undefined` and left that way. Now saved into `oldMutationObserver` and restored in the same `finally`.

Practical impact was small but real: `renderShapeCentres` is only called from this file, and each call re-sets the value, so the leak escapes only to whichever other spec files share the worker. Cheap to fix and symmetric, so there is nothing to argue about.

## `checkGlobalObject` - the finding is right, the suggested shape fails schema validation

The gap is real, and worse than the comment implies. The bare-name form catches **one** of the four ways to reach it:

```js
screen.availWidth;            // flagged
window.screen.availWidth;     // not flagged
globalThis.screen.availWidth; // not flagged
self.screen.availWidth;       // not flagged
```

So the guard I pushed was weaker than the PR body claimed. Thanks for catching that.

But `checkGlobalObject` cannot sit beside `name`. `no-restricted-globals` has an `anyOf` schema with two mutually exclusive branches (eslint 9.35.0, `lib/rules/no-restricted-globals.js:66-94`): a list of strings/`{name, message}` objects with `additionalProperties: false`, **or** a single options object `{globals, checkGlobalObject, globalObjects}` with `required: ["globals"]`. Applying the suggested diff verbatim gives:

```
Error: Key "rules": Key "no-restricted-globals":
	Unexpected property "checkGlobalObject". Expected properties: "name", "message".
	...
	Value [{...}] should match some schema in anyOf.
```

That is a hard failure at config load, so every lint run in the repo would fail, not just this rule.

The valid form nests the entry under `globals`:

```js
'no-restricted-globals': [
  'error',
  {
    globals: [{ name: 'screen', message: '...' }],
    checkGlobalObject: true,
  },
],
```

Done that way, and all four forms above are now reported. It works here specifically because `eslint.config.js` spreads `...globals.browser`, so `window` and `self` resolve in scope - `create()` silently skips any global object it cannot find in scope (`no-restricted-globals.js:218-227`), which is why a bare probe outside this repo only catches two of the four.

On the suggested spec exclusion: I have not excluded `**/*.spec.{ts,js}` from the rule. Turning it off for every spec would let a display-dependent assertion into any future test, which is the thing this rule exists to prevent. The two deliberate uses in `c4Renderer.spec.ts` carry a targeted `eslint-disable-next-line` with the reason instead - that test has to report a display size in order to prove the layout ignores it.

One note for anyone citing this option later: `checkGlobalObject` is not in the docs bundled with the package, so the implementation is the reference.

File: packages/mermaid/src/diagrams/c4/c4Renderer.spec.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    expect(await asSeenOn(400, 'c4-rows-narrow')).toEqual(await asSeenOn(4000, 'c4-rows-wide'));
+  });
+
+  it('honours a c4ShapeInRow wider than the elements would fit on screen', async () => {
+    const centres = await renderShapeCentres(
+      `C4Context\nUpdateLayoutConfig($c4ShapeInRow="6")\n${systems(6)}`,
+      'c4-rows-6'
+    );
+
+    expect(rowSizes(centres)).toEqual([6]);
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply packages/mermaid/src/diagrams/c4/c4Renderer.spec.ts:131]: Hardened, though not for the reason given - the premise about the environment is the opposite of what happens here.

**The test environment is not wide, it is 0.** jsdom hard-codes `ScreenImpl.prototype.availWidth = 0` (`node_modules/jsdom/lib/jsdom/living/window/Screen-impl.js:6`), so under vitest these tests already ran at the narrowest value possible - which is the most sensitive setting, not the least. That is verifiable rather than argued: revert `c4Renderer.ts` to `develop` and **all four** tests fail, including the two you flagged, precisely because `widthLimit` becomes 0 and every element takes its own row. If either path still consulted the display, these tests would already be catching it.

There is also a wrinkle worth naming: an earlier revision of this spec really did have the hole you describe, just via a different mechanism. The display-independence test set `availWidth` to 4000 and never cleaned up, so the two tests after it inherited 4000 and passed even against unfixed code - 2 of 4 red instead of 4 of 4. The `afterEach` removed that leak before this review. So the concern was real one revision ago, and the fix for it was ordering, not per-test width.

**Taken anyway, because the sensitivity was accidental.** It rested on jsdom's hard-coded 0; if these specs ever move to a browser environment the guarantee evaporates silently. So the width is now set explicitly, once, for the whole suite:

```ts
beforeEach(() => {
  // Report a display far narrower than a row of elements needs...
  Object.defineProperty(globalThis.screen, 'availWidth', { value: 400, configurable: true });
});
```

One `beforeEach` rather than two per-test blocks: every test in the file is then sensitive, including the two default-path ones you did not flag, and there is a single place to change if the value ever needs to. The display-independence test still overrides it per render, so it is unaffected. Confirmed all four remain red against `develop` with the narrow width in place, so the change does not weaken the evidence.

Small correction on the mechanism: the `afterEach` does not restore a saved value, it `delete`s the override so the prototype value shows through again. Right behaviour, slightly different from "restores the environment" - worth stating since it is what makes the `beforeEach` safe to add.

**Comment by filipsajdak** [comment]
Gentle nudge - open since 08-12, no review yet.

This is the determinism fix behind the question @pbrolin47 raised on #7842 during the 11.17.0 release: C4 row breaking was seeded from `screen.availWidth`, the **operating system's monitor width**. It was the only place in `packages/mermaid/src` where rendered geometry depended on the machine doing the rendering, which is why the same diagram could break rows differently in CI, in a dev server and in a user's browser - and why the release diff could not be reproduced locally.

Rows now break on `c4ShapeInRow`, so the same source always produces the same grid. There is a unit test that renders the same diagram twice with different stubbed display widths and asserts the output is identical, plus an eslint rule so `screen` cannot come back into `packages/mermaid/src`.

Two notes on the checks: the Netlify reds are stale (last build 08-12, inside the #8049 outage window; Netlify passes on newer PRs), and `argos` shows the intended row-breaking change.

No urgency, and I realise it touches a file that #8042 later deletes - the fix is deliberately small for that reason, and it is still worth having on `develop` in the meantime.

**Comment by filipsajdak** [comment]
Following my nudge a few hours ago: the Cypress-to-Playwright migration landed in between and made this conflict. Ported and mergeable again.

The conflict was in `eslint.config.js`, where this branch inserts the `no-restricted-globals` rule for `screen` directly above an entry whose globs the migration changed from `cypress/**` to `e2e/**`. Both sides kept: the rule stays, the globs take develop's form.

Two things checked after resolving, since this file is easy to break quietly: the config still loads (the options-object form of `no-restricted-globals` is schema-validated, so a bad merge here breaks every lint run), and the guard still reports `screen` when reintroduced.

The fix itself is unchanged.

**Comment by filipsajdak** [comment]
Thanks - the coverage gap is real, but I checked before writing the test and the behaviour it would assert is currently unreachable, so the test would pass no matter what `maxRowWidth()` returned.

`maxRowWidth()` is passed as the text limit for a boundary's label, type and description. All three go through:

```ts
const currentBoundaryTextWrap = currentBoundary.wrap && conf.wrap;
```

`currentBoundary.wrap` comes from `autoWrap()`, which returns `wrapEnabled` - initialised to `false`, reset to `false` in `clear()`, and only ever set by an explicit wrap directive. So the flag is false for any ordinary diagram, the text is never wrapped, and the limit has no effect on the output. That is the bug in #7949, not something this PR introduces.

I confirmed it rather than reasoning about it: rendering a boundary with a long type through the jsdom harness, with the text-measurement stub made proportional to text length so a narrow budget *would* force a break, produces a single unwrapped line. Varying `c4ShapeInRow` changes nothing about that text.

So a test of the shape you describe would be vacuous today - it would still pass if `maxRowWidth()` returned `0`, which is the opposite of a regression guard.

Two options, and I am happy with either:

1. Add a test asserting the *absence* - that boundary text does not wrap today - titled so that it fails loudly when #7949 is fixed and this becomes real behaviour. That matches how the other not-yet-supported cases are covered in the C4 characterization suite.
2. Leave it, and cover boundary wrapping properly as part of fixing #7949, where it can be asserted rather than asserted-not.

I lean towards (1) since it puts the tripwire in now, but it is a judgement call for a maintainer rather than something I want to decide unilaterally on a PR that is already waiting.

**Comment by filipsajdak** [comment]
Correcting my previous comment: I said option (1), an absence test, would work as a tripwire. I built it, and it does not - so I am not adding it, and the reason is more interesting than the test would have been.

I wrote the test (a boundary label far wider than the wrap budget, with text measured proportionally to its length, asserting it stays on one line) and then checked it was sensitive by simulating the #7949 fix: forcing `wrapEnabled` true at its declaration and in `clear()`, so `currentBoundary.wrap` comes back true. With `c4.wrap` already defaulting to true, that makes `currentBoundaryTextWrap` true.

The test still passed. A 299-character boundary label renders on a single `tspan` with wrapping fully enabled, and it stays that way with `getBBox` also made proportional rather than a fixed width. So boundary text does not wrap under any configuration I can reach, and an absence test would not fail when #7949 is fixed either - it pins nothing.

**So `maxRowWidth()` is inert**, and not only because of the flag in #7949. That is worth stating plainly for this PR: the value it replaced, `screenBounds.data.widthLimit`, was equally inert as a wrap budget before this change. This PR moves an unused number from "the operating system's monitor width" to "a function of the grid", changing what feeds it but not what it does, which is exactly what a determinism-only change should do.

The coverage gap you spotted is real. It just is not closable while the behaviour is unreachable, and I would rather say that than add a green test that asserts nothing - which is what your suggested version and my own proposed version both turn out to be. Boundary wrapping should get real assertions as part of fixing #7949, where there will be behaviour to assert.

Leaving the code as it is. Thanks for the prompt - it turned up something I had not checked.

**Comment by pbrolin47** [comment]
Hi @filipsajdak,
Sorry for the late response. A couple of hot fixes(11.17.1, 11.17.2) has been released in the last couple of days. One of them contained a fix:

https://github.com/mermaid-js/mermaid/pull/8092/changes/876e5dd747e22f05f42c915d613168e55d613d8c

which may have an impact in this PR. So I suggest to sync this PR with develop, and I'll do the review/test after that.
Regards, 
Per

**Comment by filipsajdak** [comment]
Synced - merged `develop` in at `814a24415`, clean merge, no conflicts. Thanks for the pointer to #8092; it does interact, and not in the direction I expected, so I re-measured rather than just merging.

#8092 does not touch `c4Renderer.ts` or `eslint.config.js`, so there was nothing textual to resolve. What it changes is the premise this PR's evidence rests on: with `c4LabelHelper.ts` gating on `config.c4?.wrap ?? true` again, element labels wrap to `c4.width - 2 * padding`, where before they did not wrap at all.

## The table still holds

Measured on post-#8092 `develop` and on this branch, same jsdom harness both sides, six `System(...)` elements at default config:

| `screen.availWidth` | develop, elements per row | this branch |
| --- | --- | --- |
| 0 (jsdom) | 1, 1, 1, 1, 1, 1 | 4, 2 |
| 800 (headless Chrome) | 2, 2, 2 | 4, 2 |
| 1280 (Xvfb) | 3, 3 | 4, 2 |
| 2560 (a monitor) | 4, 2 | 4, 2 |

Unchanged from the numbers in the PR body, so nothing there needs correcting.

## #8092 makes the defect purer, not smaller

The interesting part. Before the hotfix, an unwrapped label inflated `c4Shape.width` well past `c4.width`, and that width is what `Bounds.insert` compares against the display budget - so on `develop` the row structure moved with *both* the display and the label text. Now it moves with the display alone: I measured a six-element diagram whose descriptions are 100 characters each, and every box comes out at `216`, exactly `c4.width`, with identical row structure to the short-label version at every display width.

Two consequences worth stating:

- The defect is now a clean display dependency with nothing else mixed in, which is the easiest possible thing to review this against.
- `maxRowWidth()` in this PR is documented as "`c4ShapeInRow` elements at their nominal width". That assumption was false before your hotfix and is true after it, so #8092 strengthens this change rather than colliding with it.

## The one conclusion I re-checked and it survived

On 2026-08-20 I reported that `maxRowWidth()` is inert, because boundary label/type/description text never wraps. #8092 could plausibly have made that stale, since it is a change to a wrap gate. It did not: a 103-character boundary label and its type still render on a single `tspan`, and `$c4ShapeInRow="4"` versus `"2"` produces byte-identical output. The two gates are separate - #8092 changed the element-label gate in `c4LabelHelper.ts`, while boundary text is gated on `currentBoundary.wrap && conf.wrap` in `c4Renderer.ts`, which is the flag #7949 is about and is still false. So this PR still moves an unused number from the monitor width to a function of the grid.

One caveat on the measuring, since it nearly cost me a wrong answer: my first stub measured a wrapped `<text>` by its whole `textContent`, which cannot observe wrapping at all and made it look as though #8092 had changed nothing. Every number above is from a stub that measures the widest `tspan`.

## Checks

133 C4 unit tests pass, including the two `c4LabelHelper.spec.ts` cases that arrived with #8092. `eslint` and Prettier clean - worth confirming explicitly on this branch, since the `no-restricted-globals` entry is schema-validated and a bad merge there would break every lint run in the repo, not just this rule.

Ready for your review whenever suits.

**Comment by filipsajdak** [comment]
Follow-up with measured numbers, because one line in the PR description turns out to be wrong and it overstates this change.

**The description says "Argos will show a large C4 diff". It does not - Argos reports `1 changed`.** Part of that is the sheet batching (per-test screenshots are composited per spec file, so one changed sheet is not one changed diagram), but the substance is that the diff really is small, and I had not measured it before writing that sentence.

Rendering all 25 C4 fixtures under `e2e/diagrams/c4/` at several reported display widths, `develop` versus this branch:

| `screen.availWidth` | fixtures rendering differently |
| --- | --- |
| 800 (headless Chrome, the `e2e` job) | 4 of 25 |
| 1280 (Xvfb) | 1 of 25 |
| 1512 (a laptop display) | 0 of 25 |

So on a developer's machine this change is invisible on the current fixtures, and in CI it moves four of them. "Large" was wrong.

**The same measurement is the clearest statement of the defect, though.** On `develop`, these four fixtures render differently depending only on the width of the monitor:

```
characterization/elements/component-variants.mmd
characterization/elements/container-variants.mmd
characterization/elements/system-variants.mmd
characterization/relationships/directional-relationships.mmd
```

4 of 25 change between `availWidth` 800 and 1280, and another between 1280 and 1512. On this branch the count is **0 at every width** - the same source produces the same SVG whatever the display reports, which is the property the PR is for.

That is also why the row-break defect went unnoticed for so long: at typical developer display widths there is nothing to see, and it only appears where the display is narrow, which is CI.

One caveat on the numbers: they come from the jsdom harness with text measurement stubbed proportionally, so the exact count could shift by one or two in a real browser with real fonts. The 0-versus-nonzero result does not depend on that, since it is a comparison of the same harness against itself.

I will leave the description's wording to be corrected in place rather than rewriting history on an open PR - flagging it here so the claim is on the record next to the measurement.

**Comment by filipsajdak** [comment]
Gentle nudge, and no rush - following up on your 08-26 note, which I answered on 08-27 with the measured numbers (the #8092 interaction, and the correction to the description's Argos claim). Nothing has changed since; it is green apart from `argos`, which is the intended change.

This one deliberately moves screenshots, so it needs a maintainer's yes rather than a rubber stamp: rows break on `c4ShapeInRow` alone instead of being seeded from `screen.availWidth`, so a diagram lays out the same on every machine - today it gives 2 elements per row in the e2e container, 3 under Xvfb and 4 on a typical monitor. If you would rather it waited for a release boundary, that is a fine answer too; I just did not want it to look abandoned.
